k2/csrc/intersect.cu

```diff
@@ -90,950 +69,840 @@ using namespace intersect_internal;  // NOLINT
+     Intersection (a.k.a. composition) that corresponds to decoding for
      speech recognition-type tasks
 
        @param [in] a_fsas  The decoding graphs, one per sequence.  E.g. might
                            just be a linear sequence of phones, or might be
                            something more complicated.  Must have either the
-                           same Dim0() as b_fsas, or Size0()==1 in which
+                           same Dim0() as b_fsas, or Dim0() == 1 in which
                            case the graph is shared.
        @param [in] b_fsas  The neural-net output, with each frame containing the
                            log-likes of each phone.  A series of sequences of
-                           (in general) different length.
-       @param [in] beam    "Default" decoding beam.  The actual beam is dynamic
-                            and also depends on max_active and min_active.
-       @param [in] min_active  Minimum number of FSA states that are allowed to
-                           be active on any given frame for any given
-                           intersection/composition task. This is advisory,
-                           in that it will try not to have fewer than this
-                           number active.
-       @param [in] max_active  Maximum number of FSA states that are allowed to
-                           be active on any given frame for any given
-                           intersection/composition task. This is advisory,
-                           in that it will try not to exceed that but may not
-                           always succeed.
+                           (in general) different length.  MUST BE SORTED BY
+                           DECREASING LENGTH, or it is an error.
+                           (Calling code should be able to ensure this.)
+       @param [in] output_beam    Beam >0 for pruning output, i.e. arcs that are
+                           not on a path within `output_beam` of the best path
+                           will not be retained.
    */
-  MultiGraphDenseIntersect(FsaVec &a_fsas, DenseFsaVec &b_fsas,
-                           float search_beam, float output_beam,
-                           int32_t min_active, int32_t max_active)
+  MultiGraphDenseIntersect(FsaVec &a_fsas,
+                           DenseFsaVec &b_fsas,
+                           float output_beam)
       : a_fsas_(a_fsas),
         b_fsas_(b_fsas),
-        search_beam_(search_beam),
-        output_beam_(output_beam),
-        min_active_(min_active),
-        max_active_(max_active),
-        dynamic_beams_(a_fsas.Context(), b_fsas.shape.Dim0(), search_beam) {
-    NVTX_RANGE(K2_FUNC);
+        output_beam_(output_beam) {
+    NVTX_RANGE(__func__);
     c_ = GetContext(a_fsas.shape, b_fsas.shape);
+
+    K2_CHECK(a_fsas_.Dim0() == b_fsas_.shape.Dim0());
+    num_fsas_ = a_fsas_.Dim0();
+    K2_CHECK_GT(num_fsas_, 0);
     K2_CHECK(b_fsas.scores.IsContiguous());
```

Comment:
I think the constraint `IsContiguous()` is not necessary any more.

Reply:
Mm, you are right I think... should remove any mention of that from the documentation too.

Reply:
sorry I missed some of these comments from long ago (as you pointed out in another thread).. since I am away for a few days, is there any chance you could clean up these issues?
BTW, for `intersect`, it's definitely possible to improve its behavior memory-wise but this would be at the expense of time (part of what we do for speed is to make the forward and backward passes simultaneous, which is a factor of 2 in memory since otherwise we could share memory for the alphas and betas).   Let's wait until it becomes an issue and then optimize, maybe?  If it's an issue already, LMK.

One possibility is to store just every *n* frames permanently, in the forward and backward (still done simultaneously), and use temporary memory for the in-between ones; the missing ones would then be computed as needed in a loop over n.  [n might be 2 or 4 or so.]   Not a *completely* trivial change, of course.

Reply:
>  is there any chance you could clean up these issues?

working on it.